python/cuml/metrics/hinge_loss.pyx

```diff
@@ -0,0 +1,120 @@
+#
+# Copyright (c) 2021, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import cupy as cp
+from cuml.preprocessing import LabelEncoder, LabelBinarizer
+import cudf
+
+
+def cython_hinge_loss(y_true, pred_decision, labels=None, sample_weights=None):
+    """
+    Calculates non-regularized hinge loss. Adapted from scikit-learn hinge loss
+
+    Parameters
+    ----------
+    y_true: cuDF Series or cuPy array of shape (n_samples,)
+            True labels, consisting of labels for the classes.
+            In binary classification, the positive label must be
+            greater than negative class
+
+    pred_decision: cuDF DataFrame or cuPy array of shape (n_samples,) or
+                   (n_samples, n_classes)
+                   Predicted decisions, as output by decision_function (floats)
+
+    labels: cuDF Series or cuPy array, default=None
+            In multiclass problems, this must include all class labels.
+
+    sample_weight: cupy array of shape (n_samples,), default=None
+                   Sample weights to be used for computing the average
+
+    Returns
+    -------
+    loss : float.
+           The average hinge loss.
+    """
+
+    # Check types of the inputs
+    if not hasattr(y_true, "__cuda_array_interface__"):
+        raise TypeError("y_true needs to be either a cuDF Series or \
+                        a cuda_array_interface compliant array.")
+
+    if not hasattr(pred_decision, "__cuda_array_interface__") and not \
+       isinstance(pred_decision, cudf.DataFrame):
+        raise TypeError("pred_decision needs to be either a cuDF DataFrame or \
+                        a cuda_array_interface compliant array.")
```

Comment:
I would *really* avoid checking the type this way. We have helper functions in `cuml.common.input_utils` which can help with this. `determine_array_type()` would be great here.

Reply:
I'm running into this error when I include `@cuml.internals.api_return_any()`

```---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-4-153d120043af> in <module>
      1 for i in range(10):
----> 2     test_hinge_loss()

<ipython-input-3-c49bb52410db> in test_hinge_loss(input_type, nrows, ncols, n_info, datatype)
     17     cuml_model.fit(X_train, y_train)
     18     cu_predict_decision = cuml_model.decision_function(X_test)
---> 19     cu_loss = cuml_hinge(y_test, cu_predict_decision.T, labels=cp.unique(y))
     20     if input_type == "cudf":
     21         y_test = y_test.to_array()

/opt/conda/envs/rapids/lib/python3.8/site-packages/cuml/internals/api_decorators.py in inner(*args, **kwargs)
    359         def inner(*args, **kwargs):
    360             with self._recreate_cm(func, args):
--> 361                 return func(*args, **kwargs)
    362 
    363         return inner

cuml/metrics/hinge_loss.pyx in cuml.metrics.hinge_loss.cython_hinge_loss()

/opt/conda/envs/rapids/lib/python3.8/site-packages/cuml/preprocessing/LabelEncoder.py in __init__(self, handle_unknown, handle, verbose, output_type)
    131                  output_type=None):
    132 
--> 133         super().__init__(handle=handle,
    134                          verbose=verbose,
    135                          output_type=output_type)

cuml/common/base.pyx in cuml.common.base.Base.__init__()

cuml/common/base.pyx in cuml.common.base._check_output_type_str()

AssertionError: Cannot pass output_type='mirror' in Base.__init__(). Did you forget to pass `output_type=self.output_type` to a child estimator? Currently `cuml.global_output_type==`mirror`
```

Looks like the decorator is causing issues LabelEncoder call. Any suggestions on how to fix it?

Reply:
See more recent code review for a solution. Resolving